### PR TITLE
milo/support aarch64 unknown gnu

### PR DESCRIPTION
**Pull request by Milo123459**
This allows `cargo-quickbuild` to build for `aarch64-unknown-linux-gnu`.

### CODE REVIEW

**Comment by Milo123459** [comment]
cc @NobodyXu - wasn't sure how to test. this should in theory work, but needs to be tested in CI.

**Comment by Milo123459** [comment]
lmk if you have any ideas on how i can fix this because i have no idea how

File: build-version.sh
Comment (NobodyXu) [inline build-version.sh:34]: The error from the [`shellcheck`](https://github.com/cargo-bins/cargo-quickinstall/actions/runs/3915594296/jobs/6693920222) indicates that you need to quote this to prevent word splitting:

```suggestion
    wget "$(curl https://ziglang.org/download/index.json | jq 'to_entries | map([.key, .value])[1][1]["x86_64-linux"] | .tarball' | sed -e 's/^"//' -e 's/"$//')" -O zig
```

File: zig-aarch64.sh
Comment (NobodyXu) [inline zig-aarch64.sh:2]:
```suggestion
# shellcheck disable=SC2068
zig cc -target aarch64-unknown-linux-gnu $@
```

to disable the [`shellcheck`](https://github.com/cargo-bins/cargo-quickinstall/actions/runs/3915594296/jobs/6693920222#step:3:12) here.

**Review by NobodyXu — CHANGES_REQUESTED** [review]
For the linter error:

File: build-version.sh
Comment (NobodyXu) [reply build-version.sh:34]: Note that I use `-q` here to disable the progress bar, which takes up a lot of space in the CI log, makes the debugging harder.

File: build-version.sh
Comment (NobodyXu) [inline build-version.sh:43]: According to the [CI](https://github.com/cargo-bins/cargo-quickinstall/actions/runs/3915594303/jobs/6693920659#step:3:960), cargo still uses `ld`.

File: build-version.sh
Comment (NobodyXu) [inline build-version.sh:41]: I think we should instead maintain a new file `cargo-config` since using `echo` to create a file is basically unreadable.

The file should look like this according to [this SSO](https://stackoverflow.com/questions/62371337/rust-cargo-toml-specify-custom-path-for-c-linker-and-compiler):

```
[target.aarch64-unknown-linux-gnu]
linker = "./zig-aarch64.sh"
ar = "./zig-aarch64.shr"
```

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: build-version.sh
Comment (Milo123459) [reply build-version.sh:43]: weird...

File: build-version.sh
Comment (Milo123459) [reply build-version.sh:41]: hmm, i think for now it is fine.

Comment (Milo123459) [reply build-version.sh:41]: i was going off of https://actually.fyi/posts/zig-makes-rust-cross-compilation-just-work/

Comment (NobodyXu) [reply build-version.sh]: I think that's partially because:

```
echo '["linker = "./zig-aarch64.sh"]'
```

prints:

```
["linker = "./zig-aarch64.sh"]
```

which is not what cargo expected.

Comment (Milo123459) [reply build-version.sh]: I updated it in a recent commit but it's still broken

**Comment by NobodyXu** [comment]
Could you be hitting https://github.com/ziglang/zig/issues/4911 ?

**Comment by Milo123459** [comment]
I think so.

**Comment by Milo123459** [comment]
getting a new error this time. cc @NobodyXu - any ideas?

**Comment by Milo123459** [comment]
right - now im trying to make it use  the NDK version of ld. currently, when i move ld to /usr/bin, the shell doesn't pick up that there is something in that directory called `ld` and now errors out because it cannot find it. it turns out that the `sudo mv` isn't actually moving the file, i think

**Comment by Milo123459** [comment]
cc @alsuren - any ideas

File: zig-aarch64.sh
Comment (alsuren) [inline zig-aarch64.sh:3]:
```suggestion
exec zig cc -target aarch64-linux-gnu $@
```

no need to keep the shell process around in the process tree while cc runs

File: build-version.sh
Comment (alsuren) [inline build-version.sh:39]: * probably just pipe the tarball straight into tar
* can just put the zig dir first in your $PATH rather than copying stuff into /usr/local/bin?


(I'm not sure what `/usr/local/bin/lib` is supposed to be, but it feels wrong)
```suggestion
    mkdir -p zigfolder
    curl "$(curl -q https://ziglang.org/download/index.json | jq 'to_entries | map([.key, .value])[1][1]["x86_64-linux"] | .tarball' | sed -e 's/^"//' -e 's/"$//')"  | tar -x -C zigfolder --strip-components 1

    export PATH="$PWD/zigfolder:$PATH"
```

**Review by alsuren — APPROVED** [review]
Please give your squashed commit a better name, and consider my suggestions (it may be that you already tried them)

I would quite like to try one of these binaries in docker on my m1 mac. Let's ship it and get it building stuff :D

File: build-version.sh
Comment (Milo123459) [reply build-version.sh:39]: we need /lib or else it'll error with zig not being able to find some lib folder, so yeah the way it was is how we have to do it

**Comment by Milo123459** [comment]
i can't merge it because there are some review changes that haven't been addressed (even though they have). @alsuren could you merge it for me? thanks

**Comment by alsuren** [comment]
The "request changes" review is blocking the merge. I tend to avoid using that button because it feels impolite (I always tend to leave feedback as "comment"). I hadn't realised that it also blocks merging in the presence of other approvers :-(.

If you want to smoke test it before we merge to main, push it to the `actions` branch and it will kick off a one-off cronjob using your code.

**Comment by Milo123459** [comment]
alright, i'll push it! fingers crossed

**Comment by Milo123459** [comment]
seems to have worked! should i delete the `actions` branch?

**Comment by Milo123459** [comment]
nvm, it's failing because it's using a weird version of the branch causing it to not have access to a file. in practice, this won't happen if this is merged and the cron just runs.

ps @alsuren - can you not override the requested changes thing and merge it?

File: build-version.sh
Comment (alsuren) [reply build-version.sh:39]: The error you got from CI was about my broken tar command (I left out the -J)


```suggestion
    mkdir -p zigfolder
    curl "$(curl -q https://ziglang.org/download/index.json | jq 'to_entries | map([.key, .value])[1][1]["x86_64-linux"] | .tarball' | sed -e 's/^"//' -e 's/"$//')"  | tar -xJ -C zigfolder --strip-components 1

    export PATH="$PWD/zigfolder:$PATH"
```

The only reason you need to move the lib folder is that the zig binary is looking relative to itself for its lib foder. If you don't move the zig binary (and instead add it to $PATH) then I would expect this to work just fine.

**Comment by alsuren** [comment]
> ps @alsuren - can you not override the requested changes thing and merge it?

Oh hey: there is a "dismiss review" item in the menu.

I still think that https://github.com/cargo-bins/cargo-quickinstall/actions/runs/4033938309/jobs/6934774496#step:4:89  (
```
error: linker `/home/runner/work/cargo-quickinstall/cargo-quickinstall/zig-aarch64.sh` not found
```
) is a problem that we're also going to see on main. Let me have a quick dig about.

File: build-version.sh
Comment (alsuren) [inline build-version.sh:41]: I managed to make https://github.com/cargo-bins/cargo-quickinstall/actions/runs/4034113715/jobs/6935051984 pass. I think the thing that did the trick was this:

```suggestion
        echo "linker = \"$PWD/zig-aarch64.sh\"" >>~/.cargo/config
```

You can also try cherry-picking the thing that makes our selftest stuff checkout the repo in the same place as the builder does. Annoyingly it is /home/runner/work/cargo-quickinstall/cargo-quickinstall/cargo-quickinstall/ :( Might be a thing for another PR though.

**Comment by alsuren** [comment]
I just tested it out in arm docker, and `cargo-quickinstall cargo-quickinstall` now works (once I installed `curl`). Good work, and thanks for sticking with it.

Glancing at https://github.com/cargo-bins/cargo-quickinstall/pull/123/commits still doesn't tell me anything about what this PR is about, so it's likely to cause problems for anyone doing archaeology later (although they will have enough problems unpicking my commits, so they might not notice yours).

Otherwise, you have an approval and I *think* you have merge access, so ship it.

Now it's well past my bedtime. I might do some more quickinstall hacking tomorrow though.

**Comment by Milo123459** [comment]
Woo! I'll squash my commits again. Just testing the actions branch then I'll give it a merge!

File: build-version.sh
Comment (NobodyXu) [inline build-version.sh:46]: Excellent job @Milo123459 !

Looks like the issue I linked is already fixed, that's great because I also want to use zig-cc in cargo-binstall.

@Milo123459 One thing that I would you like to fix here is to use `cargo-auditable` to build the binary, just as other branches do.

You can also rewrite the other branches (the musl one) to use zig-cc and then we can also support `aarch64-unknown-linux-musl`.

**Comment by Milo123459** [comment]
Seeing some errors @alsuren 😬 https://github.com/cargo-bins/cargo-quickinstall/actions/runs/4034450545/jobs/6935671373#step:5:101

PS: I ran `cargo quickinstall cargo-binstall` on my Raspberry Pi, that has the aarch64-unknown-linux-gnu target, but a workflow run was never created for it.

**Comment by NobodyXu** [comment]
This error is caused by my PR  #140, will fix it

**Pull request merged**
